File: SimCtl.js
//var firebase = require('firebase');
//var async = require('async');

var TICKS_PER_DAY = 24 * 60;
var TICKS_PER_HOUR = 60;
var body = new HumanBody();


const EventType = {
    FOOD: 'FOOD',
    EXERCISE: 'EXERCISE',
    HALT: 'HALT',
    METFORMIN: 'METFORMIN',
    INSULIN_SHORT: 'INSULIN_SHORT',
    INSULIN_LONG: 'INSULIN_LONG'
}

class Event {
    constructor(fireTime, the_type) {
        this.fireTime_ = fireTime;
        this.eventType_ = the_type;
        this.cost0 = fireTime;
        this.cost1 = 0; // possibly redundant and can delete
    }
}

class FoodEvent extends Event {
    constructor (fireTime, quantity, foodID) {
        super(fireTime);
        this.quantity_ = quantity;
        this.foodID_ = foodID;
    }
}

class ExerciseEvent extends Event {
    constructor (fireTime, duration, exerciseID) {
        super(fireTime);
        this.duration_ = duration;
        this.exerciseID_ = exerciseID;
    }
}

class HaltEvent extends Event {
    constructor (fireTime){
        super(fireTime);
    }
}

class SimCtl {

    constructor(){
        this.ticks = 0;
        this.eventQ = new PriorityQueue();
    }

    run_simulation(){
        while(true){
            var val;
            while( (val = fire_event()) == 1);
            body.processTick();
            ticks++;
        }
    }

    fire_event(){
        var event_ = eventQ.front();

        if(!(event_)){
            console.log("No event left");
            process.exit();
        }

        if(event_.fireTime > ticks){
            return -1;
        }

        var event_type = event_.eventType_;

        switch(event_type){
            case EventType.FOOD:
                body.processFoodEvent(event_.foodID_, event_.quantity_);
                break;
            case EventType.EXERCISE:
                body.processExerciseEvent(event_.exerciseID_, event_.duration_)
                break;
            case EventType.HALT:
                process.ext();
                break;
            default:
                break;
        }

        event_ = eventQ.dequeue();
        //delete event_;
        return 1;
    }

    addEvent(fireTime, type, subtype, howmuch){
        switch (type){
            case EventType.FOOD:
                var e = new QElement(fireTime, subtype, howmuch);
                this.eventQ.enqueue(e);
                break;
            case EventType.EXERCISE:
                var e = new QElement(fireTime, subtype, howmuch);
                this.eventQ.enqueue(e);
                break;
            case EventType.HALT:
                var e = new QElement(fireTime, subtype, howmuch);
                this.eventQ.enqueue(e);
            default:
                break;               
        }
    }

    readEvents(){
        /*var rawFile = new XMLHttpRequest();
        rawFile.open("GET", file, false);
        rawFile.onreadystatechange = function(){
            if(rawFile.readyState === 4){
                if(rawFile.status === 200 || rawFile.status == 0){
                    var allText = rawFile.responseText;

                    // add stuff here!!!

                }
            }
        }
        rawFile.send(null);*/
        
        // testing
        this.addEvent(2018, EventType.FOOD, '2', 1);
        this.addEvent(2017, EventType.FOOD, '3', 1);
        this.addEvent(2016, EventType.EXERCISE, '2', 1);
        console.log(this.eventQ.front());
    }


    elapsed_days(){
        return(ticks/TICKS_PER_DAY);
    }

    elapsed_hours(){
        var x = ticks % TICKS_PER_DAY;
        return(x/TICKS_PER_HOUR);
    }

    elapsed_minutes(){
        var x = ticks % TICKS_PER_DAY;
        return(x % TICKS_PER_HOUR);
    }
}

// TODO need to figure out how to implement the main function in JavaScript
// also need to determine how we will pass in the arguments that are used when sim ran in colsole
function run_simulation(){
    while (true) {
        var val;
        while( (val = fire_event()) == 1);
        // need to convert body->processTick();
        ticks++;
    }

    return 0;
}

File: Alex/JavaSimulator-master/src/sim/SimCtl.java
package sim;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Random;

import enums.EventType;
import events.*;
import util.PriQ;

/* The global class implementing
 * the simulation controller.
 */
public class SimCtl {
	
	public static final SimCtl sim = new SimCtl();
	public static final HumanBody body = new HumanBody();
	
	public static final int TICKS_PER_DAY = 24*60;
	public static final int TICKS_PER_HOUR = 60;
	
	public static long ticks = 0;
	
    public static Random myEngine() {
    	return new Random(System.currentTimeMillis());
    }
    
    PriQ eventQ = new PriQ();
    
    public SimCtl() { ticks = 0; }
    
    public long elapsed_days() { return ticks / TICKS_PER_DAY; }
    
    public long elapsed_hours() {
    	long x = ticks % TICKS_PER_DAY;
    	return x / TICKS_PER_HOUR;
    }
    
    public long elapsed_minutes() {
    	long x = ticks % TICKS_PER_DAY;
    	return x % TICKS_PER_HOUR;
    }
    
    public static void time_stamp() {
    	System.out.print(sim.elapsed_days() + ":" + sim.elapsed_hours() + ":" + sim.elapsed_minutes() + " ");
    }
    
    public void readEvents(String filename) {
    	BufferedReader br = null;
    	FileReader fr = null;
    	try {
    		fr = new FileReader(filename);
    		br = new BufferedReader(fr);
    		String line = null;
    		while ((line = br.readLine()) != null) {
    			String[] tok = line.split(" ")[0].split(":");	// double split!
    			int day = Integer.valueOf(tok[0]);
    			int hour = Integer.valueOf(tok[1]);
    			int minutes = Integer.valueOf(tok[2]);
    			int fireTime = day * TICKS_PER_DAY + hour * TICKS_PER_HOUR + minutes;
    			
    			int type = Integer.valueOf(tok[3]);
    			int subtype = Integer.valueOf(tok[4]);
    			int howmuch = Integer.valueOf(tok[5]);
    			
    			addEvent(fireTime,type,subtype,howmuch);
    		}
    		
    	} catch (IOException e) {
    		System.out.println("Error opening file : " + filename);
    		e.printStackTrace();
    		System.exit(-1);
    	} finally {
    		try {
    			if (br != null) br.close();
    			if (fr != null) fr.close();
    		} catch (IOException ioe) { ioe.printStackTrace(); System.exit(-1); }
    	}
    }
    
    void addEvent(long fireTime, int type, long subtype, long howmuch) {
    	switch (type) {
            case 0:
                FoodEvent e = new FoodEvent(fireTime, howmuch, subtype);
                eventQ.priq_add(e);
                break;
            case 1:
                ExerciseEvent f = new ExerciseEvent(fireTime, howmuch, subtype);
                eventQ.priq_add(f);
                break;
            case 2:
                HaltEvent g = new HaltEvent(fireTime);
                eventQ.priq_add(g);
                break;
            default:
                break;
        }
    }
    
    public int fire_event() {
    	Event event_ = null;
    	try {
    		event_ = (Event) eventQ.priq_gethead();
    		if (event_.fireTime > ticks) return -1;
    	} catch (ClassCastException|NullPointerException e) {
    		System.out.println("No event left");
    		System.exit(-1);
    	}
        
        // Fire the event
        
        //System.out.println("ticks =" + ticks + ": " + elapsed_days() + "::" + elapsed_hours() + "::" + elapsed_minutes());
        //System.out.println("event_.fireTime : " + event_.fireTime);
        
        EventType event_type = event_.eventType;
        
        switch(event_type) {
            case FOOD:
                body.processFoodEvent( ((FoodEvent)event_).foodID, ((FoodEvent)event_).quantity);
                break;
            case EXERCISE:
                body.processExerciseEvent(((ExerciseEvent)event_).exerciseID, ((ExerciseEvent)event_).duration);
                break;
            case HALT:
                System.exit(0);
            default:
                break;
        } // end switch case
       	
        event_ = (Event)eventQ.priq_rmhead();
        return 1;
    }
    
    public void run_simulation() {
    	 // Always in this loop
        while(true) {
            @SuppressWarnings("unused")
			int val;
            while( (val = fire_event()) == 1 );
            // At this point:
            // no more event to fire;
            body.processTick();
            ticks++;
            //System.out.println(elapsed_days() + ":" + elapsed_hours() + ":" + elapsed_minutes());
        }
    }
    
    //friend int main(int argc, char *argv[]);
}
